Documentation for ACL V2 + ACL DRYRUN (#1756)
redis/redis-doc: topics/acl.md

*Note: an use that is not flagged with nopass, and has no list of valid passwords, is effectively impossible to use, because there will be no way to log in as such user.*

Configure selectors for the user:

> * `(<rule list>)`: Create a new selector to match rules against. Selectors are evaluated after the user permissions, and are evaluated according to the order they are defined. If a command matches either the user permissions or any selector, it is allowed.

Reviewer: add a link to the selectors section below?

    5) "commands"
    6) "-@all +get"
    7) "keys"
    8) "~cached:*"
    9) "channels"
    10) "&*"
    11) "selectors"
    12) 1) 1) "commands"

>            2) "-@all +set"

Reviewer: commenting here, i think the doc page for ACL GETUSER must mention this change.
maybe also in the history (which is part of the json file in the redis repo)

Reviewer: @madolson please make a PR to the redis-repo to add a history entry for ACL GETUSER

Author: Yes, that is why this one is still open :)

Author: https://github.com/redis/redis/pull/10200

In the previous section it was observed how it is possible to define commands
ACLs based on adding/removing single commands.

## Selectors

Redis supports additional sets of rules that are evaluated independently of the main user permissions.
These secondary sets of permissions are called selectors and added by wrapping a set of rules within parentheses.
Selectors are evaluated after the main user permissions are checked and are evaluated in the order they are added to the user. 

> A command is allowed if either the root user permissions or any selectors matches a command.

Reviewer 2: ```suggestion
A command is allowed if either the main user permissions or any selectors matches a command.
```

Author: Not sure about this, seems like I should stick with one way to describe the root user permission. I would think I should update everything to main in this case.

For example, consider a user with the ACL rules `+GET ~key1 (+SET ~key2)`.
This user is able to execute `GET key1` and `SET key2 hello`, but not `GET key2` or `SET key1 world`.

> Unlike the root user permissions, selectors can not be modified after they are added.
> Selectors can be removed with the `clearselectors` keyword, which removes all of the added selectors.
> Note that `clearselectors` does not remove the root user permissions.

Reviewer 2: ```suggestion
Unlike the main user permissions, selectors cannot be modified after they are added.
Instead, selectors can be removed with the `clearselectors` keyword, which removes all of the added selectors.
Note that `clearselectors` does not remove the main user permissions.
```

## Key permissions

> Redis key patterns can also be used to define how a command is able to touch a key.

Reviewer: ```suggestion
Starting with Redis 7.0, Redis key patterns can also be used to define how a command is able to touch a key.
```

This is achieved through rules that define key permissions.
The key permission rules take the form of `%(<permission>)~<pattern>`.
Permissions are defined as individual characters that map to the following key permissions:

* W (Write): The data stored within the key may be updated or deleted. 
* R (Read): User supplied data from the key is processed, copied or returned. Note that this does not include metadata such as size information (example `STRLEN`), type information (example `TYPE`) or information about whether a value exists within a collection (example `SISMEMBER`). 

Permissions can be composed together by specifying multiple characters. 
Specifying the permission as 'RW' is considered full access and is analogous to just passing in `~<pattern>`.

For a concrete example, consider a user with ACL rules `+@all ~app1:* (+@readonly ~app2:*)`.
This user has full access on `app1:*` and readonly access on `app2:*`.
However, some commands support reading data from one key, doing some transformation, and storing it into another key.
One such command is the `COPY` command, which copies the data from the source key into the destination key.

> The example set of ACL rules is unable to handle a request copying data from `app2:user` into `app1:user`, since neither the root permission or the selector fully matches the command.

Reviewer 2: root permission => main rule set?

However, using key selectors you can define a set of ACL rules that can handle this request `+@all ~app1:* %R~app2:*`.
The first pattern is able to match `app1:user` and the second pattern is able to match `app2:user`.

> Which type of permission is required for a command is documented through [key specifications](/topics/key-specs#logical-operation-flags).

Reviewer: i think there's a way to provide a direct link, maybe there's a better way than `https://redis.io/topics/key-specs#logical-operation-flags`. @itamarhaber

Reviewer: ohh, never mind. i see this is the right syntax.

The type of permission is based off the keys logical operation flags. 
The insert, update, and delete flags map to the write key permission. 

> The access flag flag maps to the read key permission.

Reviewer 2: ```suggestion
The access flag maps to the read key permission.
```

Note: Side channels are ignored when it comes to evaluating read permissions.
Many write commands return metadata about the modified key, for example `LPUSH` returns the size of the resulting list.

> Considering any metadata to be a read would limit the utility of the write key permissions.

Reviewer 2: ```suggestion
Considering any metadata to be a read would limit the utility of the read key permissions.
```

???

Reviewer 2: Oh, right, i rememver this discussion in the original PR. Perhaps it can be explained even more thoroughly or with emphasis on *write* to highlight that it's not a typo.

Author: Yeah, I think my mistake was I start by talking about read, then I pivot to talking about write. In reality it's just a conversation about read/lack of read permissions.
